### ISSUE TITLE
[Bug] DFlash2 acceptance=0 on V100 TP4 (Qwen3.8-27B + DFlash2)

### ISSUE THREAD

**Issue by wdxfwq** [1CatAI/1Cat-vLLM-0.0.1#478]
### Your current environment

1Cat-vLLM v1.5.0 (wheel 1cat_vllm-1.5.0-cp312)4x Tesla V100-SXM2-32GB (SM70), dual Xeon E5-2696 v4CUDA 12.8, PyTorch 2.10 (cu128), Python 3.12, Ubuntu 22.04Target: RadixArk/Qwen3.8-27B-NVFP4-BF16-LMHeadDraft: incoai/Qwen3.8-27B-DFlash2

### 🐛 Describe the bug

DFlash2 speculative decoding has zero acceptance on 4xV100 TP4. Draft tokens are generated but all rejected by the target (spec_decode_num_accepted_tokens_total stays 0.0).## Serve command```vllm serve /path/to/Qwen3.8-27B-NVFP4-BF16-LMHead \  --served-model-name qwen3.8-27b-dflash2 --trust-remote-code \  --tensor-parallel-size 4 --attention-backend FLASH_ATTN_V100 \  --kv-cache-dtype fp8_e5m2 --max-model-len 8192 \  --gpu-memory-utilization 0.80 --language-model-only \  --reasoning-parser qwen3 \  --speculative-config '{"method":"dflash","model":"/path/to/Qwen3.8-27B-DFlash2","kv_cache_dtype":"auto"}'```## ExpectedAcceptance length ~4-5 (README documents 251.60 tok/s, AL 4.686, PR #288).## Actual- spec_decode_num_draft_tokens_total = hundreds- spec_decode_num_accepted_tokens_total = 0.0 (all positions rejected)- Target-only decode works: ~47 tok/s, correct output## Tried (all still yield acceptance=0)1. QUASAR-QAT (QAT) target -> acceptance=0; RadixArk BF16-LMHead (PTQ) -> acceptance=02. --language-model-only -> multimodal warning gone, acceptance still 03. Draft kv_cache_dtype fp8->float16 -> still 04. --max-num-batched-tokens 4096 -> still 05. Draft attention TRITON_ATTN -> rejected ("non-causal attention not supported")## Root cause observationThe DFlash2 draft has no embedding layer (81 tensors, no embed_tokens); it must consume the target's hidden states. The warning "does not support external multimodal embeddings ... using text-only draft inputs instead" suggests the target->drafter hidden-state handoff fails on the V100/SM70 TP4 path.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by wdxfwq** [comment]
**Root cause narrowed down: an SM70 BF16-emulation scale bug in the DFlash2 draft model.**

I used the `VLLM_SPEC_DUMP_ALIGNMENT=1` diagnostics (from #291) to dump the selector alignment on decode steps, then cross-referenced the draft candidates against the target's top-20 logits. Full evidence chain below — I think this is enough to pinpoint the bug.

### 1. Draft candidates are garbage when BF16 emulation is ON (the default)

With the draft config at its shipped `dtype: bfloat16` (which forces `_use_sm70_bf16_emulation = True` on SM70, since there is no env-var override), the draft candidate tokens have almost zero overlap with the target top-20:

- Row-by-row overlap rate: **2.86%**
- Target top-20 decodes to `=` `-th` `:` `is` `must` `should` `int` `can` (semantically correct code tokens)
- Draft candidates decode to `фа` (Cyrillic) / `专利技术` `显眼` (Chinese) / `_REPLY` / `oglob` / `umus` (cross-language garbage)
- Global intersection of the two vocab-ID sets: 1201 ∩ 817 = only **48** tokens

### 2. Disabling BF16 emulation fixes alignment → the bug is in the SM70 emulation itself

I temporarily flipped the draft config to `dtype: float16` so `_use_sm70_bf16_emulation` returns `False`. Overlap rate jumped from 2.86% → **90.5%** (152/168 rows).

But this is *not* a fix: the draft's `unary_logits` / `lattice_scores` / `realized_logits` then become **all NaN** (draft weights are stored bfloat16 and overflow/underflow in float16). So the BF16 emulation is *required* on SM70 — it must be fixed, not bypassed.

### 3. Already ruled out (all verified correct)

- **QPN8 rerank** — `VLLM_SM70_DFLASH2_QPN8_RERANK_SHADOW=1` reports `support_missing=0`, `exact_set_rows=all`, `top1_match_rows=all`, `ordered_value_max_abs≈0` → identical to the dense path.
- **fc sharding** — `VLLM_SM70_DFLASH2_SHARDED_CONTEXT_FC=0` → draft candidates byte-identical.
- **lm_head sharing** — verified `SHARED lm_head (draft->target) is_target=True`; the draft correctly reuses the target's valid lm_head (target weight std=0.0140, absmax=0.2656 vs the draft's `torch.empty` garbage).

### 4. Likely bug location: scale-convention mismatch in the BF16 emulation

In `vllm/model_executor/models/dflash_sm70.py` the emulation uses `DFLASH_SM70_WIDE_OUTPUT_SCALE = 256.0`, and the decoder-layer RMSNorm runs with `residual_scale=256.0`. But in `qwen3_dflash2.py`:

- `hidden_norm` (the context norm after the fc projection) uses `residual_scale=1.0`;
- the shipped draft config has **no** `input_embedding_scale` / `output_multiplier`, so both default to `1.0`.

So the first decoder layer's residual stream (token embedding at standard scale ≈1.0) is mixed with the later layers' WIDE scale (256.0) — a scale mismatch that would corrupt the draft hidden states exactly as the dump shows.

### Environment

1Cat-vLLM v1.5.0, 4× Tesla V100-SXM2-32GB (SM70), CUDA 12.8 / PyTorch 2.10 (cu128) / Python 3.12 / Ubuntu 22.04. Target `dfischermittwald/Qwen3.8-27B-NVFP4-DFlash2`, draft `z-lab/Qwen3.8-27B-DFlash2` (dtype bfloat16). Serve has `--language-model-only` (the earlier modality-detection fix) already applied — that fix is correct and necessary, but this second bug still yields acceptance=0.
